src/token/feature-ext-parser.js

```diff
@@ -0,0 +1,79 @@
+const async = require('async');
+
+const FedAuthInfoID = {
+  STSURL: 0x01,
+  SPN: 0x02
+};
+
+const TERMINATOR = 0xFF;
+
+const infoData = {
+  spnLen: undefined,
+  stsurlLen: undefined
+};
+
+var readFedAuthInfoOpt = (parser, callback) => {
+  parser.readUInt8((fedauthInfoID) => {
+    parser.readUInt32LE((fedAuthInfoDataLen) => {
+      parser.readUInt32LE((fedAuthInfoDataOffset) => {
+
+        switch (fedauthInfoID) {
+          case FedAuthInfoID.SPN:
+            infoData.spnLen = fedAuthInfoDataLen ;
+            break;
+          case FedAuthInfoID.STSURL:
+            infoData.stsurlLen = fedAuthInfoDataLen ;
+            break;
+            // ignoring unknown fedauthinfo options
+          default:
+            break;
+        }
+        if (infoData.spnLen && infoData.stsurlLen) {
+          parser.readBuffer(infoData.spnLen, (spn) => {
+            parser.readBuffer(infoData.stsurlLen, (stsurl) => {
+              callback({
+                'name': 'FEDAUTHINFO',
+                'event': 'fedAuthInfo',
+                'fedAuthInfoData': {
+                  stsurl: stsurl.toString().replace(/\0/g, ''),
+                  spn: spn.toString().replace(/\0/g, '')
+                }
+              });
+            });
+          });
+        }
+      });
+    });
+  });
+};
+
+module.exports.fedAuthInfoParser = fedAuthInfoParser;
+function fedAuthInfoParser(parser, colMetadata, options, callback) {
+  parser.readUInt32LE((tokenLength) => {
+    parser.readUInt32LE((countOfInfoIDs) => {
```

Comment:
Nesting many `parser.readXYZ` calls is really expensive and slow, due to callback nesting involved. 😞

Could we change this to instead just read `tokenLength` via `parser.readUInt32LE`, and then read the whole token into a `Buffer` in one go via `parser.readBuffer(tokenLength, (data) => { ... })`. All other data extraction could then happen using regular `Buffer` operations, without any callbacks being involved.

@Hadis-Fard What do you think? 😄

Reply:
good idea, I play with it to see what i can do